src/Nethermind/Nethermind.Blockchain/BlockTree.cs

```diff
@@ -1170,37 +1136,29 @@ private void UpdateHeadBlock(Block block)
 
         private void UpdateOrCreateLevel(long number, BlockInfo blockInfo)
         {
-            ChainLevelInfo level = LoadLevel(number, false);
-
-            if (level != null)
+            using (var batch = _blockInfoRepository.StartBatch())
             {
-                BlockInfo[] blockInfos = new BlockInfo[level.BlockInfos.Length + 1];
-                for (int i = 0; i < level.BlockInfos.Length; i++)
+                ChainLevelInfo level = LoadLevel(number, false);
+
+                if (level != null)
                 {
-                    blockInfos[i] = level.BlockInfos[i];
+                    BlockInfo[] blockInfos = level.BlockInfos;
+                    Array.Resize(ref blockInfos, blockInfos.Length + 1);
+                    blockInfos[blockInfos.Length - 1] = blockInfo;
+                    level.BlockInfos = blockInfos;
```

Comment:
it is by ref then do you need this assignment?

Reply:
I mean this is most likely not needed: level.BlockInfos = blockInfos;

Reply:
also the level state is corrupted for a moment here so if you have multithreaded access it may lead to bad behaviour

Reply:
Because its ref to local variable and it can change.
It would have to be:  Array.Resize(ref level.BlockInfos, blockInfos.Length + 1); But it won't compile